### PR TITLE
Add Control Drum Related Functionalities to Reactor Module #19100

### LINKED ISSUE #19100: Adding Control Drum Related Functionalities to Reactor Module
## Reason
A new MOOSE module called `Reactor` was recently created to contain objects that enable intrinsic generation of meshes with complex geometries involved in reactor cores (e.g., hexagonal-geometry reactor cores) as well as related utilities to facilitate their simulations. While the initial implementation was focused on the fundamental pin cell, assembly, and core hexagonal-geometry meshing capabilities, this issue aims to add control drum related objects.

## Design
A new `Mesh` object will be added to enable meshing of a steady state snapshot of a control drum by modifying the mesh generated by `PolygonConcentricCircleMeshGenerator` or `HexagonConcentricCircleAdaptiveBoundaryMeshGenerator`.

In addition, a new `Function` object will be added to enable modeling of time-dependent control drum rotation. This object can assign real-time values that measure and report absorber volume fraction in rotating control drum ring regions. The assigned values can then be used by different physical models to appropriately mix materials and simulate rotating control drums in a reactor core.

## Impact
With the features described in this issue, MOOSE users will be able to create meshes for static control drums for steady state simulations as well as quantitatively describe time-dependent control drum rotation for transient simulations when those control drums are contained within hexagonal geometries.

### LINKED ISSUE #19285: Add methods for getting multiple MeshGenerators
## Reason
We have `getMesh()` and `getMeshByName()`. There should be methods for obtaining multiple generators.

## Design
Add `getMeshes()` and `getMeshesByName()`.

## Impact
Simplicity. Also allows for more const.

### PR DESCRIPTION
**Pull request by shikhar413**
Closes #19100 and closes #19285

### CODE REVIEW

**Comment by miaoyinb** [comment]
@loganharbour This is the new PR that adds the control drum related stuffs. Thanks.

PS: @shikhar413 will handle submission of this PR and future PRs of the `MOOSE Meshing Improvement project`. It would be great if he could be added to the automatic testing list.

**Review by loganharbour — CHANGES_REQUESTED** [review]
Here's a start. Still need to look at docs and tests.

**Review by GiudGiud — CHANGES_REQUESTED** [review]

**Review by GiudGiud — CHANGES_REQUESTED** [review]
Can you please add to moose/modules/doc/content/citing.md for the reactor module?

**Review by GiudGiud — DISMISSED** [review]
In the future when renaming an object / a file, please consider using `git mv` as this reduces the size of the diffs.
Nbd here since the renamed file is really tiny. 
@loganharbour want to have a second look?

**Review by loganharbour — DISMISSED** [review]

**Review by loganharbour — CHANGES_REQUESTED** [review]
I missed one last thing - can you please move the image content into the `large_media` submodule? Feel free to contact me on Slack or over e-mail if you need assistance with this.

I'm looking into the distributed mesh implementation right now.

**Comment by miaoyinb** [comment]
@loganharbour Please let us know if anything needs to be done on our side to solve the conflicts issue. Thanks.

**Comment by loganharbour** [comment]
@miaoyinb Replicated mesh requirements are in place. I did a lot of cleanup in 10426e8 and ca52a82. Please take a look when you have a change. I think I figured out the distributed issue.

@shikhar413 I'm getting there :-)

**Comment by miaoyinb** [comment]
> @miaoyinb Replicated mesh requirements are in place. I did a lot of cleanup in [10426e8](https://github.com/idaholab/moose/commit/10426e8a6f86f638c49de5df336ef8f3df11522f) and [ca52a82](https://github.com/idaholab/moose/commit/ca52a826d94dd87819f4c862ae452c7e82bd3251). I removed the passing of the current node ID in most of the generation in the polygon generator because it was redundant (libmesh will keep track of this); I can add this back if needed. Let me know. Please take a look when you have a chance. I think I figured out the distributed issue.
> 
> @shikhar413 I'm getting there :-)

Thanks a lot @loganharbour. I will take a look,

**Comment by miaoyinb** [comment]
@loganharbour Thank you so much for the cleanup work 👍 and figuring out the distributed issue. It looks great now. I will also apply the approaches you used to my future codes. Let us know if anything else needs to be done by us before moving forward. Thanks again.

**Comment by loganharbour** [comment]
I'm going to defer the distributed recover issue for `HexagonConcentricCircleAdaptiveBoundaryMeshGenerator`. It's driving me into the wall, and I don't want to hold this back anymore. Skipping is okay for now.

blue_crab patch is almost done for the changes I made to mesh generators.

**Review by loganharbour — DISMISSED** [review]
Approving content that is not my own.

@GiudGiud I'll merge once you approve and I have the blue_crab patch ready.

**Comment by miaoyinb** [comment]
> I'm going to defer the distributed recover issue for `HexagonConcentricCircleAdaptiveBoundaryMeshGenerator`. It's driving me into the wall, and I don't want to hold this back anymore. Skipping is okay for now.
> 
> blue_crab patch is almost done for the changes I made to mesh generators.

Thank you, @loganharbour

File: modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:218]: ??

Comment (loganharbour) [reply modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:218]: grep has failed me

Comment (loganharbour) [inline modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:218]:
```suggestion
   * @param subdomain_rings numbers of radial intervals of all involved subdomain layers
```

File: modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h
Comment (miaoyinb) [inline modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:320]: seems to be another one.

Comment (loganharbour) [inline modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:320]:
```suggestion
   * @param pitch pitch size of the involved hexagon mesh
```

File: modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:416]: do you need this declared ahead of time

**Review by GiudGiud — APPROVED** [review]

**Comment by miaoyinb** [comment]
@GiudGiud @loganharbour Thank you so much for your review and help on this PR!

**Pull request merged**
